README.md
# awespottr
> Find the cheapest EC2 spot prices across all regions

[![npmjs.com Version](https://img.shields.io/npm/v/awespottr.svg?maxAge=2592000)](https://www.npmjs.com/package/awespottr)

## Usage

When installed globally with `npm install -g awespottr`, you can run
awespottr on the command line:

- Get help: `awespottr -h`

- Get spot prices for a g2.8xlarge instance: `awespottr g2.8xlarge`

- Specify an AWS CLI profile (credentials):
`AWS_PROFILE=awespottr awespottr m4.large`

- Limit results to a specific region: `awespottr --region us-west-2 c4.xlarge`

## IAM Policy

awespottr must be run with an IAM role or user with permissions to get
information about AWS regions, availability zones, and spot price history.

This IAM policy contains the minimum permissions required for running awespottr:

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Stmt1470583998000",
            "Effect": "Allow",
            "Action": [
                "ec2:DescribeAvailabilityZones",
                "ec2:DescribeRegions",
                "ec2:DescribeSpotPriceHistory"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
```

## Example Output

Running `awespottr m4.2xlarge` should produce output like:

```
Checking spot prices for m4.2xlarge instance type.

AWS Zone                 Hourly Rate
------------------------ ------------
eu-west-1b               $0.063000
eu-central-1a            $0.067000
eu-central-1b            $0.072700
ap-northeast-2c          $0.073300
ap-southeast-2b          $0.078100
us-west-2c               $0.078100
ap-southeast-1a          $0.079800
ap-southeast-1b          $0.081200
eu-west-1c               $0.086200
ap-northeast-1a          $0.086600
us-east-1e               $0.088900
ap-northeast-1c          $0.090200
us-west-1c               $0.091600
us-west-1b               $0.096700
ap-northeast-2a          $0.096800
eu-west-1a               $0.099200
ap-southeast-2a          $0.099600
ap-southeast-2c          $0.103100
us-east-1a               $0.178500
ap-south-1a              $0.292000
us-west-2b               $0.390700
ap-south-1b              $0.430000
us-east-1c               $0.479000
us-west-2a               $1.000000
us-east-1d               $1.500000

Cheapest hourly rate for m4.2xlarge is $0.063000 in zone eu-west-1b
```


CHANGELOG.md
# Change Log for awespottr

## v0.4.0 [2019/03/06]

- Get multi-instance search working
- Add -n flag to limit output to top n cheapest options

## v0.3.0 [2018/05/26]

- Fixing numeric comparison logic used to identify cheapest availability zone.

## v0.2.0 [2016/08/15]

- Adding command line option for filtering by region.
- Allowing awespottr to be installed globally as an executable.
- Documentation improvements.

## v0.1.0 [2016/08/08]

- Initial release of awespottr.


cli.js
#!/usr/bin/env node

var _ = require('lodash');
var AWS = require('aws-sdk');
var chalk = require('chalk');
var moment = require('moment');
var program = require('commander');

var ec2Type = [];
var awsRegions = {};
var zones = {};
var lowPrice = false;
var lowZone = false;

function getRegions() {
  return new Promise(function(resolve, reject) {
    var ec2 = new AWS.EC2({apiVersion: '2016-04-01', region: 'us-west-2'});
    ec2.describeRegions({}, function(err, data) {
      if (err) {
        console.log(err, err.stack);
        return reject(err);
      }

      var promises = [];
      if (data && data.Regions) {
        _.each(data.Regions, function(region) {
          if (region.RegionName && !awsRegions.hasOwnProperty(region.RegionName)) {
            awsRegions[region.RegionName] = {};
            promises.push(getRegionSpots(region.RegionName));
          }
        })
      }
      return Promise.all(promises).then(resolve);
    });
  });
}

function getRegionSpots(regionName) {
  return new Promise(function(resolve, reject) {
    var ec2 = new AWS.EC2({apiVersion: '2016-04-01', region: regionName});
    var params = {
      InstanceTypes: ec2Type,
      ProductDescriptions: ['Linux/UNIX', 'Linux/UNIX (Amazon VPC)'],
      StartTime: moment().subtract(4, 'hours').utc().toDate()
    };
    ec2.describeSpotPriceHistory(params, function(err, data) {
      if (err) {
        console.log(err, err.stack);
        return reject(err);
      }

      if (data && data.SpotPriceHistory) {
        _.each(data.SpotPriceHistory, function(sdata) {
          if (sdata.SpotPrice && sdata.Timestamp && sdata.AvailabilityZone) {
            var sdate = moment(sdata.Timestamp);
            if (!zones.hasOwnProperty(sdata.AvailabilityZone) || zones[sdata.AvailabilityZone].lastDate < sdate) {
              zones[sdata.InstanceType + " " + sdata.AvailabilityZone] = {
                lastDate: sdate,
                lastPrice: sdata.SpotPrice,
                zone: sdata.AvailabilityZone,
                itype: sdata.InstanceType
              };
              if (!lowPrice || Number(sdata.SpotPrice) < lowPrice) {
                lowPrice = Number(sdata.SpotPrice);
                lowZone = sdata.AvailabilityZone;
              }
            }
          }
        });
      }
      resolve();
    });
  });
}

function handleResults() {
  console.log('\n' + _.padEnd('Instance Type', 16) + ' ' + _.padEnd('AWS Zone', 24) + ' ' + _.padEnd('Hourly Rate', 12));
  console.log(_.pad('', 16, '-') + ' ' + _.pad('', 24, '-') + ' ' + _.pad('', 12, '-'));
  var sortedZones = _.values(zones);
  sortedZones = _.sortBy(sortedZones, function(val) { return Number(val.lastPrice); }, 'zone');
  if (program.number > 0) {
    sortedZones = sortedZones.slice(0, program.number)
  }
  _.each(sortedZones, function(data) {
    var msg = _.padEnd(data.itype, 16) + ' ' + _.padEnd(data.zone, 24) + ' ' + _.padEnd('$' + data.lastPrice, 12);
    if (data.zone === lowZone && Number(data.lastPrice) === Number(lowPrice)) {
      msg = chalk.green(msg);
    } else if (lowPrice * 1.1 >= data.lastPrice) {
      msg = chalk.yellow(msg);
    }
    console.log(msg);
  });
  console.log(chalk.green('\nCheapest hourly rate for [' + ec2Type.join(', ') + '] is $' + lowPrice + ' in zone ' + lowZone));
}

exports.standalone = function (cb) {
  ec2Type = [];
  awsRegions = {};
  zones = {};
  lowPrice = false;
  lowZone = false;
  program
    .version('0.4.0')
    .usage('[options] <EC2 instance types ...>')
    .option('-r, --region <AWS region>', 'Limit to the given region')
    .option('-n, --number <number to show>', 'Only show the top few cheapest spots')
    .action(function(type, othertypes) {
      ec2Type = []
      ec2Type.push(type)
      if (typeof othertypes === 'string' || othertypes instanceof String) {
        ec2Type.push(othertypes)
      }
      if (Array.isArray(othertypes)) {
        ec2Type.concat(othertypes)
      }
    })
    .parse(process.argv);

  console.log('Checking spot prices for [' + ec2Type.join(', ') + '] instance type(s).');

  if (program.region) {
    console.log('Limiting results to region ' + program.region);
    getRegionSpots(program.region)
      .then(handleResults)
      .then(cb)
      .catch(function(err) {
        console.log('error', err);
      });
  } else {
    getRegions(program.region)
      .then(handleResults)
      .then(cb)
      .catch(function(err) {
        console.log('error', err);
      });
  }
}

if (!module.parent) {
  exports.standalone()
}


test-cli.js
'use strict'
/* global describe it */
const chai = require('chai')
chai.use(require('chai-diff'))
chai.config.showDiff = true
const expect = chai.expect
const sinon = require('sinon')
const AWS = require('aws-sdk')
const cli = require('./cli')

const RESULTS = {
  'eu-north-1': {
    'm5.metal': [
      { AvailabilityZone: 'eu-north-1c',
        InstanceType: 'm5.metal',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '0.917300',
        Timestamp: '2019-03-06T00:54:47.000Z' },
      { AvailabilityZone: 'eu-north-1b',
        InstanceType: 'm5.metal',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '0.917300',
        Timestamp: '2019-03-06T00:54:47.000Z' },
      { AvailabilityZone: 'eu-north-1a',
        InstanceType: 'm5.metal',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '0.917300',
        Timestamp: '2019-03-06T00:54:47.000Z' }
    ],
    'm5.24xlarge': [
      { AvailabilityZone: 'eu-north-1a',
        InstanceType: 'm5.24xlarge',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '0.957800',
        Timestamp: '2019-03-06T00:54:47.000Z' },
      { AvailabilityZone: 'eu-north-1b',
        InstanceType: 'm5.24xlarge',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.457100',
        Timestamp: '2019-03-06T00:54:47.000Z' },
      { AvailabilityZone: 'eu-north-1c',
        InstanceType: 'm5.24xlarge',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.457100',
        Timestamp: '2019-03-06T00:54:47.000Z' }
      ]
  },
  'us-east-1': {
    'm5.metal': [
      { AvailabilityZone: 'us-east-1a',
        InstanceType: 'm5.metal',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.088200',
        Timestamp: '2019-03-05T22:05:47.000Z' },
      { AvailabilityZone: 'us-east-1f',
        InstanceType: 'm5.metal',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.088200',
        Timestamp: '2019-03-05T22:05:47.000Z' },
      { AvailabilityZone: 'us-east-1d',
        InstanceType: 'm5.metal',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.088200',
        Timestamp: '2019-03-05T22:05:47.000Z' }
    ],
    'm5.24xlarge': [
      { AvailabilityZone: 'us-east-1a',
        InstanceType: 'm5.24xlarge',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '0.957800',
        Timestamp: '2019-03-06T00:54:47.000Z' },
      { AvailabilityZone: 'us-east-1b',
        InstanceType: 'm5.24xlarge',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.457100',
        Timestamp: '2019-03-06T00:54:47.000Z' },
      { AvailabilityZone: 'us-east-1c',
        InstanceType: 'm5.24xlarge',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.457100',
        Timestamp: '2019-03-06T00:54:47.000Z' }
    ]
  },
  'us-west-2': {
    'm5.metal': [
      { AvailabilityZone: 'us-west-2c',
        InstanceType: 'm5.metal',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.081800',
        Timestamp: '2019-03-05T21:40:30.000Z' },
      { AvailabilityZone: 'us-west-2a',
        InstanceType: 'm5.metal',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.081800',
        Timestamp: '2019-03-05T21:40:30.000Z' },
      { AvailabilityZone: 'us-west-2b',
        InstanceType: 'm5.metal',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.081800',
        Timestamp: '2019-03-05T21:40:30.000Z' }
    ],
    'm5.24xlarge': [
      { AvailabilityZone: 'us-west-2a',
        InstanceType: 'm5.24xlarge',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '0.957800',
        Timestamp: '2019-03-06T00:54:47.000Z' },
      { AvailabilityZone: 'us-west-2b',
        InstanceType: 'm5.24xlarge',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.457100',
        Timestamp: '2019-03-06T00:54:47.000Z' },
      { AvailabilityZone: 'us-west-2c',
        InstanceType: 'm5.24xlarge',
        ProductDescription: 'Linux/UNIX',
        SpotPrice: '1.457100',
        Timestamp: '2019-03-06T00:54:47.000Z' }
    ]

  }
}

class EC2Stub {
  constructor (params) {
    this.region = params.region
  }
  
  describeRegions(params = {}, callback) {
    callback(null, {
      Regions: [
        { Endpoint: 'ec2.eu-north-1.amazonaws.com',
          RegionName: 'eu-north-1' },
        { Endpoint: 'ec2.us-east-1.amazonaws.com',
          RegionName: 'us-east-1' },
        { Endpoint: 'ec2.us-west-2.amazonaws.com',
          RegionName: 'us-west-2' }
      ]
    })
  }
  describeSpotPriceHistory(params = {}, callback) {
    const results = {
      NextToken: '',
      SpotPriceHistory: []
    }
    for (let itype of params.InstanceTypes) {
      if (RESULTS[this.region][itype]) {
        results.SpotPriceHistory = results.SpotPriceHistory.concat(RESULTS[this.region][itype])
      }
    }
    callback(null, results)
  }
}

describe('Awespottr', function () {
  beforeEach( function (done) {
    this.sandbox = sinon.createSandbox()
    this.sandbox.stub(AWS, 'EC2').callsFake(args => new EC2Stub(args))
    const orig = console.log
    this.consoleout = ''
    this.sandbox.stub(console, 'log').callsFake(a => {
      this.consoleout += a + "\n"
    })
    done()
  })
  it('does an all-region lookup, single instance type', function (done) {
    process.argv = ['node', 'cli.js', 'm5.metal']
    cli.standalone(() => {
      this.sandbox.restore()
      expect(this.consoleout).not.to.be.differentFrom(`Checking spot prices for [m5.metal] instance type(s).

Instance Type    AWS Zone                 Hourly Rate 
---------------- ------------------------ ------------
\u001b[33mm5.metal         eu-north-1a              $0.917300   \u001b[39m
\u001b[33mm5.metal         eu-north-1b              $0.917300   \u001b[39m
\u001b[32mm5.metal         eu-north-1c              $0.917300   \u001b[39m
m5.metal         us-west-2a               $1.081800   
m5.metal         us-west-2b               $1.081800   
m5.metal         us-west-2c               $1.081800   
m5.metal         us-east-1a               $1.088200   
m5.metal         us-east-1d               $1.088200   
m5.metal         us-east-1f               $1.088200   
\u001b[32m
Cheapest hourly rate for [m5.metal] is $0.9173 in zone eu-north-1c\u001b[39m
`)
      done()
    })
  })
  it('does a all-region lookup, multiple instance types', function (done) {
    process.argv = ['node', 'cli.js', 'm5.metal', 'm5.24xlarge', 'm0.fakebox']
    cli.standalone(() => {
      this.sandbox.restore()
      expect(this.consoleout).not.to.be.differentFrom(`Checking spot prices for [m5.metal, m5.24xlarge] instance type(s).

Instance Type    AWS Zone                 Hourly Rate 
---------------- ------------------------ ------------
\u001b[33mm5.metal         eu-north-1a              $0.917300   \u001b[39m
\u001b[33mm5.metal         eu-north-1b              $0.917300   \u001b[39m
\u001b[32mm5.metal         eu-north-1c              $0.917300   \u001b[39m
\u001b[33mm5.24xlarge      eu-north-1a              $0.957800   \u001b[39m
\u001b[33mm5.24xlarge      us-east-1a               $0.957800   \u001b[39m
\u001b[33mm5.24xlarge      us-west-2a               $0.957800   \u001b[39m
m5.metal         us-west-2a               $1.081800   
m5.metal         us-west-2b               $1.081800   
m5.metal         us-west-2c               $1.081800   
m5.metal         us-east-1a               $1.088200   
m5.metal         us-east-1d               $1.088200   
m5.metal         us-east-1f               $1.088200   
m5.24xlarge      eu-north-1b              $1.457100   
m5.24xlarge      eu-north-1c              $1.457100   
m5.24xlarge      us-east-1b               $1.457100   
m5.24xlarge      us-east-1c               $1.457100   
m5.24xlarge      us-west-2b               $1.457100   
m5.24xlarge      us-west-2c               $1.457100   
\u001b[32m
Cheapest hourly rate for [m5.metal, m5.24xlarge] is $0.9173 in zone eu-north-1c\u001b[39m
`)
      done()
    })
  })
  it('does a single-region lookup, single instance type', function (done) {
    process.argv = ['node', 'cli.js', '-r', 'eu-north-1', 'm5.metal']
    cli.standalone(() => {
      this.sandbox.restore()
      expect(this.consoleout).not.to.be.differentFrom(`Checking spot prices for [m5.metal] instance type(s).
Limiting results to region eu-north-1

Instance Type    AWS Zone                 Hourly Rate 
---------------- ------------------------ ------------
\u001b[33mm5.metal         eu-north-1a              $0.917300   \u001b[39m
\u001b[33mm5.metal         eu-north-1b              $0.917300   \u001b[39m
\u001b[32mm5.metal         eu-north-1c              $0.917300   \u001b[39m
\u001b[32m
Cheapest hourly rate for [m5.metal] is $0.9173 in zone eu-north-1c\u001b[39m
`)
      done()
    })
  })
  it('does a single-region lookup, multiple instance types', function (done) {
    process.argv = ['node', 'cli.js', '-r', 'eu-north-1', 'm5.metal', 'm5.24xlarge', 'm0.fakebox']
    cli.standalone(() => {
      this.sandbox.restore()
      expect(this.consoleout).not.to.be.differentFrom(`Checking spot prices for [m5.metal, m5.24xlarge] instance type(s).
Limiting results to region eu-north-1

Instance Type    AWS Zone                 Hourly Rate 
---------------- ------------------------ ------------
\u001b[33mm5.metal         eu-north-1a              $0.917300   \u001b[39m
\u001b[33mm5.metal         eu-north-1b              $0.917300   \u001b[39m
\u001b[32mm5.metal         eu-north-1c              $0.917300   \u001b[39m
\u001b[33mm5.24xlarge      eu-north-1a              $0.957800   \u001b[39m
m5.24xlarge      eu-north-1b              $1.457100   
m5.24xlarge      eu-north-1c              $1.457100   
\u001b[32m
Cheapest hourly rate for [m5.metal, m5.24xlarge] is $0.9173 in zone eu-north-1c\u001b[39m
`)
      done()
    })
  })
  it('does a single-region lookup, multiple instance types, limited to top 3', function (done) {
    process.argv = ['node', 'cli.js', '-n', '3', '-r', 'eu-north-1', 'm5.metal', 'm5.24xlarge', 'm0.fakebox']
    cli.standalone(() => {
      this.sandbox.restore()
      expect(this.consoleout).not.to.be.differentFrom(`Checking spot prices for [m5.metal, m5.24xlarge] instance type(s).
Limiting results to region eu-north-1

Instance Type    AWS Zone                 Hourly Rate 
---------------- ------------------------ ------------
\u001b[33mm5.metal         eu-north-1a              $0.917300   \u001b[39m
\u001b[33mm5.metal         eu-north-1b              $0.917300   \u001b[39m
\u001b[32mm5.metal         eu-north-1c              $0.917300   \u001b[39m
\u001b[32m
Cheapest hourly rate for [m5.metal, m5.24xlarge] is $0.9173 in zone eu-north-1c\u001b[39m
`)
      done()
    })
  })
})
